src/window/basic_frame.rs

```diff
@@ -413,56 +416,59 @@ impl Frame for BasicFrame {
-                            - ((ROUNDING_SIZE as f32).powi(2)
-                                - ((ROUNDING_SIZE - y) as f32).powi(2)).sqrt()
-                                as u32;
-
-                        for x in 0..width {
-                            if x >= circle_width && x < width - circle_width {
-                                let _ = writer.write(color);
-                            } else {
-                                let _ = writer.write(&[0x00, 0x00, 0x00, 0x00]);
-                            }
-                        }
-                    } else {
-                        for _ in 0..width {
-                            let _ = writer.write(color);
-                        }
-                    }
+                let mut mmap = pool.mmap();
+                {
+                    let mut header_canvas = Canvas::new(
+                        &mut mmap[0..HEADER_SIZE as usize * width as usize * 4],
+                        width as usize,
+                        HEADER_SIZE as usize,
+                        width as usize * 4,
+                    );
+                    header_canvas.clear();
+
+                    let header_bar = rectangle::Rectangle::new(
+                        (0, 0),
+                        (width as usize - 1, HEADER_SIZE as usize - 1),
+                        Some((
+                            HEADER_SIZE as usize,
+                            color,
+                            rectangle::Sides::TOP,
+                            Some(ROUNDING_SIZE as usize),
+                        )),
+                        None,
+                    );
+                    header_canvas.draw(&header_bar);
+
+                    draw_buttons(
+                        &mut header_canvas,
+                        width,
+                        true,
+                        &self
+                            .pointers
+                            .iter()
+                            .flat_map(|p| {
+                                if p.is_alive() {
+                                    let data: &Mutex<
+                                        PointerUserData,
+                                    > = p.user_data().unwrap();
+                                    Some(data.lock().unwrap().location)
+                                } else {
+                                    None
+                                }
+                            }).collect(),
+                    );
                 }
 
-                // For every pixel in borders
-                for _ in BORDER_SIZE * width..pxcount {
-                    let _ = writer.write(&[0x00, 0x00, 0x00, 0x00]);
+                // For each pixel in borders
+                {
+                    let mut writer = &mut mmap[HEADER_SIZE as usize * width as usize * 4..];
+                    for _ in 0..pxcount {
+                        let _ = writer.write(&[0x00, 0x00, 0x00, 0x00]);
+                    }
```

Comment:
hmm, if `writer` is of type `&mut [u8]`, I don't think this code will fill it with zeros, given the `io::Write` implementation for `&mut [u8]` cannot track advancement in the buffer. This likely just "works" because the default value for the content when growing a file is already 0.

If think this can be replaced by

```rust
for b in &mut mmap[HEADER_SIZE as usize * width as usize * 4..] {
    *b = 0x00;
}
```